packages/transaction-metrics-exporter/src/block-processor.ts

```
@@ -0,0 +1,266 @@
+
+    // Prevent same block multiples times
+    let lastBlocks: number[] = []
+    subscription.on('data', (header) => {
+      if (!lastBlocks.includes(header.number)) {
+        // tslint:disable-next-line: no-floating-promises
+        this.onNewBlock(header.number)
+      }
+      lastBlocks.push(header.number)
+      lastBlocks = lastBlocks.slice(-10)
+    })
+  }
+
+  async initBatch() {
+    let block = this.fromBlock - 1
+    while (++block <= (await this.getToBlock())) {
+      await this.onNewBlock(block, block - this.fromBlock)
+    }
+  }
+
+  async getToBlock() {
+    if (this.toBlock) {
+      return this.toBlock
+    }
+    return (await this.kit.web3.eth.getBlock('latest')).number
+  }
+
+  async onNewBlock(blockNumber: number, blockIntervalNumber?: number) {
+    if ((blockIntervalNumber ?? blockNumber) % this.blockInterval === 0) {
+      await Promise.all([this.fetchBlockState(blockNumber), this.processBlockHeader(blockNumber)])
+    }
+  }
+
+  async fetchBlockState(blockNumber: number) {
+    const promises = stateGetters.map(
+      async ({ contract, method, args, transformValues, maxBucketSize }) => {
+        if (!(await this.contracts[contract].exists(blockNumber))) {
+          return
+        }
+        args = typeof args === 'function' ? args(blockNumber) : args
+        this.contracts[contract].setDefaultBlock(blockNumber)
+        return (this.contracts as any)[contract][method](...args)
+          .then((returnData: any) => {
+            this.contracts[contract].setDefaultBlock('latest')
+            const data: DataResult = {
+              contract,
+              function: method,
+              args: JSON.stringify(args),
+              blockNumber,
+              values: transformValues(returnData),
+            }
+            this.logEvent(LoggingCategory.State, data)
+            this.logHistogram({ ...data, args }, maxBucketSize)
+          })
+          .catch(() => '') as Promise<any>
```

I don't think we should be just failing silently?